{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/soumyabroto/soumyabroto/blob/main/GeneGPT.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "B6UQtybRlGjm"
      },
      "outputs": [],
      "source": [
        "!pip install -q -U google-generativeai"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "Q3ddFxBlKbMh",
        "outputId": "c43e3bfb-30fe-4e17-cea0-f9f67bdf097a"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Cloning into 'GeneGPT'...\n",
            "remote: Enumerating objects: 203, done.\u001b[K\n",
            "remote: Counting objects: 100% (203/203), done.\u001b[K\n",
            "remote: Compressing objects: 100% (72/72), done.\u001b[K\n",
            "remote: Total 203 (delta 145), reused 168 (delta 127), pack-reused 0\u001b[K\n",
            "Receiving objects: 100% (203/203), 1.87 MiB | 9.46 MiB/s, done.\n",
            "Resolving deltas: 100% (145/145), done.\n"
          ]
        }
      ],
      "source": [
        "!git clone https://github.com/ncbi/GeneGPT"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "kk9Va4GU5-4m",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 36
        },
        "outputId": "3b0e81fc-4fdf-45fd-b301-7f723cf80b06"
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "'/content'"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "string"
            }
          },
          "metadata": {},
          "execution_count": 2
        }
      ],
      "source": [
        "pwd"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "02AYYoHgqJoK",
        "outputId": "762a8b32-b2b3-4c42-dbf7-6824a77ad521"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "[Errno 2] No such file or directory: '/content/GeneGPT'\n",
            "/content\n"
          ]
        }
      ],
      "source": [
        "cd /content/GeneGPT"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 35
        },
        "id": "ZCQOQUtiqKHo",
        "outputId": "9658f5e2-e19c-40c7-dec5-d1a5f52062f2"
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "'/content/GeneGPT'"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "string"
            }
          },
          "metadata": {},
          "execution_count": 3
        }
      ],
      "source": [
        "pwd"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "lI27qYiypwZl",
        "outputId": "cc462fd5-760a-456d-cd72-6f4be8a3e7ac"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\u001b[31mERROR: Could not open requirements file: [Errno 2] No such file or directory: 'requirements.txt'\u001b[0m\u001b[31m\n",
            "\u001b[0m"
          ]
        }
      ],
      "source": [
        "!pip install -r requirements.txt\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "V4e7d_kvqQAR"
      },
      "outputs": [],
      "source": []
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "PjBJQVvbjBh6"
      },
      "outputs": [],
      "source": [
        "# from google.colab import userdata\n",
        "\n",
        "# # Or use `os.getenv('GOOGLE_API_KEY')` to fetch an environment variable.\n",
        "# GOOGLE_API_KEY=userdata.get('GOOGLE_API_KEY')\n",
        "\n",
        "# genai.configure(api_key=GOOGLE_API_KEY)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "JH7F4iEJdXv-"
      },
      "outputs": [],
      "source": [
        "import os\n",
        "\n",
        "os.environ[\"GOOGLE_API_KEY\"] = \"AIzaSyDtWSVt17MUPAMF7wG0l_XzX8o-7bNi0Eg\""
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "SUmh2y6TiJ-C"
      },
      "outputs": [],
      "source": [
        "# for m in genai.list_models():\n",
        "#   if 'generateContent' in m.supported_generation_methods:\n",
        "#     print(m.name)"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "generation_config = {\n",
        "  \"temperature\": 0.2,\n",
        "  \"top_p\": 1,\n",
        "  \"top_k\": 32,\n",
        "}"
      ],
      "metadata": {
        "id": "uJeRo59lZBGq"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "hKcXPvl1iJ7D",
        "outputId": "5cd0bfdf-42ea-4582-cedf-50630ad4b47f"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "The Moon is Earth's only natural satellite.  It's a celestial body that orbits our planet and is the fifth-largest moon in the Solar System.  Here's a breakdown of key characteristics:\n",
            "\n",
            "* **Formation:**  The prevailing theory is that the Moon formed from debris ejected after a Mars-sized object collided with Earth billions of years ago. This debris coalesced in orbit around our planet.\n",
            "\n",
            "* **Composition:**  The Moon is made up primarily of rock and metal, similar to Earth, but with a slightly different composition.  Its surface is covered in a layer of dust and debris called regolith.\n",
            "\n",
            "* **Orbit and Rotation:** The Moon takes approximately 27.3 days to orbit Earth and also 27.3 days to rotate once on its axis. This synchronous rotation is why we always see the same side of the Moon from Earth.\n",
            "\n",
            "* **Phases:**  As the Moon orbits Earth, the portion illuminated by the Sun changes, creating the lunar phases we observe (new moon, crescent moon, half moon, full moon, etc.).\n",
            "\n",
            "* **Influence on Earth:** The Moon's gravity is the primary cause of Earth's tides.  It also plays a role in stabilizing Earth's axial tilt, which contributes to our relatively stable climate.\n",
            "\n",
            "* **Exploration:**  The Moon has been the target of numerous robotic missions and crewed landings, most famously the Apollo missions, which landed the first humans on the Moon in 1969.\n",
            "\n",
            "* **Future Plans:**  There are ongoing plans for future lunar missions, including the establishment of a permanent human presence on the Moon as a stepping stone for further space exploration.\n",
            "\n"
          ]
        }
      ],
      "source": [
        "import google.generativeai as genai\n",
        "from google.generativeai.types import HarmCategory, HarmBlockThreshold\n",
        "\n",
        "model = genai.GenerativeModel(model_name='gemini-1.5-pro', generation_config=generation_config)\n",
        "response = model.generate_content(\"What is moon?\")\n",
        "\n",
        "print(response.text)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "wgHtiCyl00uN",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "36116a40-afc3-4dc2-c7e8-2f2d7f51028a"
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "genai.GenerativeModel(\n",
              "    model_name='models/gemini-1.5-pro',\n",
              "    generation_config={'temperature': 0.2, 'top_p': 1, 'top_k': 32},\n",
              "    safety_settings={},\n",
              "    tools=None,\n",
              "    system_instruction=None,\n",
              "    cached_content=None\n",
              ")"
            ]
          },
          "metadata": {},
          "execution_count": 9
        }
      ],
      "source": [
        "model"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "5cKd4H0Old40"
      },
      "outputs": [],
      "source": []
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "Qc-Tu7tIaaVg"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "roZCesZxrUeR"
      },
      "outputs": [],
      "source": [
        "import json\n",
        "# import openai\n",
        "# import config\n",
        "# openai.api_key = config.API_KEY\n",
        "\n",
        "import os\n",
        "import re\n",
        "import sys\n",
        "import time\n",
        "import urllib.request\n",
        "\n",
        "def call_api(url):\n",
        "\ttime.sleep(1)\n",
        "\turl = url.replace(' ', '+')\n",
        "\tprint(url)\n",
        "\n",
        "\treq = urllib.request.Request(url)\n",
        "\twith urllib.request.urlopen(req) as response:\n",
        "\t\tcall = response.read()\n",
        "\n",
        "\treturn call\n",
        "\n",
        "def get_prompt_header(mask):\n",
        "\t'''\n",
        "\tmask: [1/0 x 6], denotes whether each prompt component is used\n",
        "\n",
        "\toutput: prompt\n",
        "\t'''\n",
        "\turl_1 = 'https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi?db=gene&retmax=5&retmode=json&sort=relevance&term=LMP10'\n",
        "\tcall_1 = call_api(url_1)\n",
        "\n",
        "\turl_2 = 'https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=gene&retmax=5&retmode=json&id=19171,5699,8138'\n",
        "\tcall_2 = call_api(url_2)\n",
        "\n",
        "\turl_3 = 'https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esummary.fcgi?db=snp&retmax=10&retmode=json&id=1217074595'\n",
        "\tcall_3 = call_api(url_3)\n",
        "\n",
        "\turl_4 = 'https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi?db=Pubmed&retmax=20&retmode=json&sort=relevance&term=Meesmann+corneal+dystrophy'\n",
        "\tcall_4 = call_api(url_4)\n",
        "\n",
        "\turl_5 = 'https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esummary.fcgi?db=omim&retmax=20&retmode=json&id=618767,601687,300778,148043,122100'\n",
        "\tcall_5 = call_api(url_5)\n",
        "\n",
        "\turl_6 = 'https://blast.ncbi.nlm.nih.gov/blast/Blast.cgi?CMD=Put&PROGRAM=blastn&MEGABLAST=on&DATABASE=nt&FORMAT_TYPE=XML&QUERY=ATTCTGCCTTTAGTAATTTGATGACAGAGACTTCTTGGGAACCACAGCCAGGGAGCCACCCTTTACTCCACCAACAGGTGGCTTATATCCAATCTGAGAAAGAAAGAAAAAAAAAAAAGTATTTCTCT&HITLIST_SIZE=5'\n",
        "\tcall_6 = call_api(url_6)\n",
        "\trid = re.search('RID = (.*)\\n', call_6.decode('utf-8')).group(1)\n",
        "\n",
        "\turl_7 = f'https://blast.ncbi.nlm.nih.gov/blast/Blast.cgi?CMD=Get&FORMAT_TYPE=Text&RID={rid}'\n",
        "\ttime.sleep(30)\n",
        "\tcall_7 = call_api(url_7)\n",
        "\n",
        "\tprompt = ''\n",
        "\tprompt += 'Hello. Your task is to use NCBI Web APIs to answer genomic questions.\\n'\n",
        "\t#prompt += 'There are two types of Web APIs you can use: Eutils and BLAST.\\n\\n'\n",
        "\n",
        "\tif mask[0]:\n",
        "\t\t# Doc 0 is about Eutils\n",
        "\t\tprompt += 'You can call Eutils by: \"[https://eutils.ncbi.nlm.nih.gov/entrez/eutils/{esearch|efetch|esummary}.fcgi?rettype=abstract&retmode=text&db=Pubmed&retmax={}&{term|id}={term|id}]\".\\n'\n",
        "\t\tprompt += 'esearch: input is a search term and output is database id(s).\\n'\n",
        "\t\tprompt += 'efectch/esummary: input is database id(s) and output is full records or summaries that contain name, chromosome location, and other information.\\n'\n",
        "\t\tprompt += 'Normally, you need to first call esearch to get the database id(s) of the search term, and then call efectch/esummary to get the information with the database id(s).\\n'\n",
        "\t\t# prompt += 'Database: gene is for genes, snp is for SNPs, and omim is for genetic diseases.\\n\\n'\n",
        "\tif mask[1]:\n",
        "\t\t# Doc 1 is about BLAST\n",
        "\t\tprompt += 'For DNA sequences, you can use BLAST by: \"[https://blast.ncbi.nlm.nih.gov/blast/Blast.cgi?CMD={Put|Get}&PROGRAM=blastn&MEGABLAST=on&DATABASE=nt&FORMAT_TYPE={XML|Text}&QUERY={sequence}&HITLIST_SIZE={max_hit_size}]\".\\n'\n",
        "\t\tprompt += 'BLAST maps a specific DNA {sequence} to its chromosome location among different specices.\\n'\n",
        "\t\tprompt += 'You need to first PUT the BLAST request and then GET the results using the RID returned by PUT.\\n\\n'\n",
        "\n",
        "\tif any(mask[2:]):\n",
        "\t\tprompt += 'Here are some examples:\\n\\n'\n",
        "\n",
        "\tif mask[2]:\n",
        "\t\t# Example 1 is from gene alias task\n",
        "\t\tprompt += f'Question: What is the official gene symbol of LMP10?\\n'\n",
        "\t\tprompt += f'[{url_1}]->[{call_1}]\\n'\n",
        "\t\tprompt += f'[{url_2}]->[{call_2}]\\n'\n",
        "\t\tprompt += f'Answer: PSMB10\\n\\n'\n",
        "\n",
        "\tif mask[3]:\n",
        "\t\t# Example 2 is from SNP gene task\n",
        "\t\tprompt += f'Question: Which gene is SNP rs1217074595 associated with?\\n'\n",
        "\t\tprompt += f'[{url_3}]->[{call_3}]\\n'\n",
        "\t\tprompt += f'Answer: LINC01270\\n\\n'\n",
        "\n",
        "\tif mask[4]:\n",
        "\t\t# Example 3 is from gene disease association\n",
        "\t\tprompt += f'Question: What are genes related to Meesmann corneal dystrophy?\\n'\n",
        "\t\tprompt += f'[{url_4}]->[{call_4}]\\n'\n",
        "\t\tprompt += f'[{url_5}]->[{call_5}]\\n'\n",
        "\t\tprompt += f'Answer: KRT12, KRT3\\n\\n'\n",
        "\n",
        "\tif mask[5]:\n",
        "\t\t# Example 4 is for BLAST\n",
        "\t\tprompt += f'Question: Align the DNA sequence to the human genome:ATTCTGCCTTTAGTAATTTGATGACAGAGACTTCTTGGGAACCACAGCCAGGGAGCCACCCTTTACTCCACCAACAGGTGGCTTATATCCAATCTGAGAAAGAAAGAAAAAAAAAAAAGTATTTCTCT\\n'\n",
        "\t\tprompt += f'[{url_6}]->[{rid}]\\n'\n",
        "\t\tprompt += f'[{url_7}]->[{call_7}]\\n'\n",
        "\t\tprompt += f'Answer: chr15:91950805-91950932\\n\\n'\n",
        "\n",
        "\treturn prompt"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "bu8YUAbChAaP"
      },
      "outputs": [],
      "source": []
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "qG-q1saS9Bz1",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "65ce84e7-ca30-436d-ce32-1182a13dc482"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi?db=gene&retmax=5&retmode=json&sort=relevance&term=LMP10\n",
            "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=gene&retmax=5&retmode=json&id=19171,5699,8138\n",
            "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esummary.fcgi?db=snp&retmax=10&retmode=json&id=1217074595\n",
            "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi?db=Pubmed&retmax=20&retmode=json&sort=relevance&term=Meesmann+corneal+dystrophy\n",
            "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esummary.fcgi?db=omim&retmax=20&retmode=json&id=618767,601687,300778,148043,122100\n",
            "https://blast.ncbi.nlm.nih.gov/blast/Blast.cgi?CMD=Put&PROGRAM=blastn&MEGABLAST=on&DATABASE=nt&FORMAT_TYPE=XML&QUERY=ATTCTGCCTTTAGTAATTTGATGACAGAGACTTCTTGGGAACCACAGCCAGGGAGCCACCCTTTACTCCACCAACAGGTGGCTTATATCCAATCTGAGAAAGAAAGAAAAAAAAAAAAGTATTTCTCT&HITLIST_SIZE=5\n",
            "https://blast.ncbi.nlm.nih.gov/blast/Blast.cgi?CMD=Get&FORMAT_TYPE=Text&RID=NABHGRSW013\n",
            "where does ids3 chromosome occur?\n"
          ]
        }
      ],
      "source": [
        "cut_length = 18000\n",
        "\n",
        "# str_mask is a string of six 0/1 marking whether a in-context learning component is used\n",
        "# six digits correspond to Dc. 1-2, Dm. 1-4\n",
        "\n",
        "# str_mask = sys.argv[1]\n",
        "# str_mask = \"111111\"\n",
        "str_mask = \"100000\"\n",
        "mask = [bool(int(x)) for x in str_mask]\n",
        "prompt = get_prompt_header(mask)\n",
        "\n",
        "# results are saved in the dir of six digits\n",
        "if not os.path.isdir(str_mask):\n",
        "  os.mkdir(str_mask)\n",
        "\n",
        "\n",
        "question = \"where does ids3 chromosome occur?\"\n",
        "print(question)\n",
        "q_prompt = prompt + f'Question: {question}\\n'\n",
        "\n",
        "# save the prompting logs\n",
        "prompts = []\n",
        "\n",
        "# record API call times\n",
        "num_calls = 0"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "print(q_prompt)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "Golz2otkH61E",
        "outputId": "b2896e7d-5a90-4f00-a5a6-185dc692a862"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Hello. Your task is to use NCBI Web APIs to answer genomic questions.\n",
            "You can call Eutils by: \"[https://eutils.ncbi.nlm.nih.gov/entrez/eutils/{esearch|efetch|esummary}.fcgi?rettype=abstract&retmode=text&db=Pubmed&retmax={}&{term|id}={term|id}]\".\n",
            "esearch: input is a search term and output is database id(s).\n",
            "efectch/esummary: input is database id(s) and output is full records or summaries that contain name, chromosome location, and other information.\n",
            "Normally, you need to first call esearch to get the database id(s) of the search term, and then call efectch/esummary to get the information with the database id(s).\n",
            "Question: where does ids3 chromosome occur?\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "5ASnuUWSsU-w"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "0m4F-zCDIAxt"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "UKI3eUZnNTse"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "03NsgpILba8t",
        "outputId": "5a12266b-b00a-4574-9e7e-81c10280a7eb"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "User Prompt: Hello. Your task is to use NCBI Web APIs to answer genomic questions.\n",
            "You can call Eutils by: \"[https://eutils.ncbi.nlm.nih.gov/entrez/eutils/{esearch|efetch|esummary}.fcgi?rettype=abstract&retmode=text&db=Pubmed&retmax={}&{term|id}={term|id}]\".\n",
            "esearch: input is a search term and output is database id(s).\n",
            "efectch/esummary: input is database id(s) and output is full records or summaries that contain name, chromosome location, and other information.\n",
            "Normally, you need to first call esearch to get the database id(s) of the search term, and then call efectch/esummary to get the information with the database id(s).\n",
            "Question: where does ids3 chromosome occur?\n",
            "\n",
            "--------------------------------------------------\n"
          ]
        }
      ],
      "source": [
        "cut_length = 18000\n",
        "\n",
        "# str_mask is a string of six 0/1 marking whether a in-context learning component is used\n",
        "# six digits correspond to Dc. 1-2, Dm. 1-4\n",
        "\n",
        "# # str_mask = sys.argv[1]\n",
        "# str_mask = \"111111\"\n",
        "# mask = [bool(int(x)) for x in str_mask]\n",
        "# prompt = get_prompt_header(mask)\n",
        "\n",
        "# # results are saved in the dir of six digits\n",
        "# if not os.path.isdir(str_mask):\n",
        "#   os.mkdir(str_mask)\n",
        "\n",
        "\n",
        "# question = \"What is IDS3?\"\n",
        "# print(question)\n",
        "# q_prompt = prompt + f'Question: {question}\\n'\n",
        "\n",
        "# save the prompting logs\n",
        "prompts = []\n",
        "\n",
        "# record API call times\n",
        "num_calls = 0\n",
        "\n",
        "if True:\n",
        "  if len(q_prompt) > cut_length:\n",
        "    # truncate from the start\n",
        "    q_prompt = q_prompt[len(q_prompt) - cut_length:]\n",
        "\n",
        "  # delta = time.time() - prev_call\n",
        "\n",
        "  # # codex has a rate limite of 20 requests / min\n",
        "  # # it's a workaround\n",
        "  # if delta < 3.1:\n",
        "  #   time.sleep(3.1 - delta)\n",
        "\n",
        "  try:\n",
        "    prev_call = time.time()\n",
        "    # response = openai.Completion.create(**body)\n",
        "    print(f\"User Prompt: {q_prompt}\\n{'-'*50}\")\n",
        "    response = model.generate_content(q_prompt + \"If it is a question like what is IDS3 or where does IDS3 occur, always use Pubmed db. Always set retmode=json in the links\",\n",
        "                                      safety_settings = {\n",
        "                                          HarmCategory.HARM_CATEGORY_HATE_SPEECH: HarmBlockThreshold.BLOCK_NONE,\n",
        "                                          HarmCategory.HARM_CATEGORY_HARASSMENT: HarmBlockThreshold.BLOCK_NONE,\n",
        "                                          HarmCategory.HARM_CATEGORY_SEXUALLY_EXPLICIT: HarmBlockThreshold.BLOCK_NONE,\n",
        "                                          HarmCategory.HARM_CATEGORY_DANGEROUS_CONTENT: HarmBlockThreshold.BLOCK_NONE,\n",
        "                                      })\n",
        "  # except openai.error.InvalidRequestError:\n",
        "  except:\n",
        "    # output.append([question, answer, 'lengthError', prompts])\n",
        "    # break\n",
        "    pass\n",
        "\n",
        "  # text = response['choices'][0]['text']\n",
        "  text = response.text\n",
        "  # print(text)"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "print(text)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "-7eQmsYZwZO2",
        "outputId": "c226f48f-80a3-4389-93a9-eb30386bfedc"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "1. **Search for \"IDS3\" in PubMed:**\n",
            "\n",
            "[https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi?retmode=json&db=pubmed&term=IDS3](https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi?retmode=json&db=pubmed&term=IDS3)\n",
            "\n",
            "This search will return a JSON object containing the PubMed IDs (PMIDs) related to IDS3.  We need to extract the `IdList` from the result.  Let's assume the `IdList` contains  `[12345678, 98765432, ...]`.\n",
            "\n",
            "\n",
            "2. **Fetch summaries for the identified PMIDs:**\n",
            "\n",
            "We'll use the esummary utility to get information about these PMIDs.  We can request multiple IDs at once by separating them with commas.  Let's use the hypothetical IDs from the previous step:\n",
            "\n",
            "[https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esummary.fcgi?retmode=json&db=pubmed&id=12345678,98765432](https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esummary.fcgi?retmode=json&db=pubmed&id=12345678,98765432)\n",
            "\n",
            "This will return a JSON object containing summaries of the publications.  We need to examine these summaries to find information about the chromosomal location of IDS3.  Specifically, look for mentions of chromosome number within the `result` object.\n",
            "\n",
            "**Important Considerations:**\n",
            "\n",
            "* **Real-world scenario:**  You'll need to programmatically parse the JSON responses to extract the `IdList` and then use it in the `esummary` call.\n",
            "* **No guarantee of location information:**  PubMed summaries may not always contain explicit chromosomal location information. You might need to resort to reading the full articles (which is more complex and beyond the scope of simple API calls).  You could try adding \"chromosome\" to your search term (e.g., \"IDS3 chromosome\") to potentially narrow down the results to articles more likely to contain this information.\n",
            "* **Gene databases:** For precise gene location information, it's generally more reliable to use dedicated gene databases like Gene (accessible via Entrez as well) rather than relying on PubMed.  A search in Gene would be a more direct approach.\n",
            "\n",
            "\n",
            "**Example using Gene database:**\n",
            "\n",
            "1. **Search for \"IDS3\" in Gene:**\n",
            "\n",
            "[https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi?retmode=json&db=gene&term=IDS3](https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi?retmode=json&db=gene&term=IDS3)\n",
            "\n",
            "2. **Fetch summaries for the identified Gene IDs:**  Similar to the PubMed example, extract the `IdList` and use `esummary` with the `gene` database. The summaries from Gene are much more likely to contain precise chromosomal location information.\n",
            "\n",
            "\n",
            "By using the Gene database and parsing the `esummary` results, you should be able to reliably determine the chromosomal location of IDS3.\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "71zY-1TBIDG2"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "n6hDiWtrba38",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "bfa34d80-9139-40af-9bdd-f458557262b2"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "['https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi?retmode=json&db=pubmed&term=IDS3', 'https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esummary.fcgi?retmode=json&db=pubmed&id=12345678,98765432', 'https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi?retmode=json&db=gene&term=IDS3']\n",
            "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi?retmode=json&db=pubmed&term=IDS3\n",
            "b'{\"header\":{\"type\":\"esearch\",\"version\":\"0.3\"},\"esearchresult\":{\"count\":\"17\",\"retmax\":\"17\",\"retstart\":\"0\",\"idlist\":[\"39448407\",\"33771198\",\"31681346\",\"29882727\",\"29666780\",\"29664328\",\"28359213\",\"28298243\",\"24351075\",\"23675379\",\"17478636\",\"17467282\",\"16887895\",\"11346963\",\"11117263\",\"11070970\",\"8019781\"],\"translationset\":[],\"querytranslation\":\"\\\\\"IDS3\\\\\"[All Fields]\"}}\\n'\n"
          ]
        }
      ],
      "source": [
        "url_regex = r'\\[(https?://[^\\[\\]]+)\\]'\n",
        "matches = re.findall(url_regex, text)\n",
        "print(matches)\n",
        "\n",
        "if matches:\n",
        "  url = matches[0]\n",
        "  # print(url)\n",
        "\n",
        "  # wait till the BLAST is done on NCBI server\n",
        "  if 'blast' in url and 'Get' in url: time.sleep(30)\n",
        "\n",
        "  call = call_api(url)\n",
        "\n",
        "  if 'blast' in url and 'Put' in url:\n",
        "    rid = re.search('RID = (.*)\\n', call.decode('utf-8')).group(1)\n",
        "    call = rid\n",
        "\n",
        "  if len(call) > 10000:\n",
        "    call = call[:10000]\n",
        "\n",
        "  q_prompt = f'{q_prompt}{text}->[{call}]\\n'\n",
        "\n",
        "print(call)"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "matches"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "SLKc52SJLsD9",
        "outputId": "1301eda4-ec4a-4dfd-c418-690cbe4dd062"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "['https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi?retmode=json&db=pubmed&term=IDS3',\n",
              " 'https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esummary.fcgi?retmode=json&db=pubmed&id=12345678,98765432',\n",
              " 'https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi?retmode=json&db=gene&term=IDS3']"
            ]
          },
          "metadata": {},
          "execution_count": 16
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "print(url)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "VABGWRMXLodl",
        "outputId": "15970d9f-02f5-4dd2-9b88-6f479f5d3995"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi?retmode=json&db=pubmed&term=IDS3\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# call_api(\"https://www.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi?db=Pubmed&retmax=5&retmode=json&sort=relevance&term=IDS3\")"
      ],
      "metadata": {
        "id": "hSIgeZCMwnJ8"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "call"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "adMNQh8iOT5z",
        "outputId": "2f6de114-3b12-47ad-e7b9-147aaf2d3995"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "b'{\"header\":{\"type\":\"esearch\",\"version\":\"0.3\"},\"esearchresult\":{\"count\":\"17\",\"retmax\":\"17\",\"retstart\":\"0\",\"idlist\":[\"39448407\",\"33771198\",\"31681346\",\"29882727\",\"29666780\",\"29664328\",\"28359213\",\"28298243\",\"24351075\",\"23675379\",\"17478636\",\"17467282\",\"16887895\",\"11346963\",\"11117263\",\"11070970\",\"8019781\"],\"translationset\":[],\"querytranslation\":\"\\\\\"IDS3\\\\\"[All Fields]\"}}\\n'"
            ]
          },
          "metadata": {},
          "execution_count": 18
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "r7_N_c0Sa-Fc"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "QU1a9HZ9ba1d"
      },
      "outputs": [],
      "source": [
        "json_string = call.decode('utf-8')\n",
        "\n",
        "# Convert JSON string to dictionary\n",
        "data = json.loads(json_string)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "TUp3OzSlbaxm",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "8429e3df-93f9-4569-94e7-7b95a33e1d71"
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "['39448407',\n",
              " '33771198',\n",
              " '31681346',\n",
              " '29882727',\n",
              " '29666780',\n",
              " '29664328',\n",
              " '28359213',\n",
              " '28298243',\n",
              " '24351075',\n",
              " '23675379',\n",
              " '17478636',\n",
              " '17467282',\n",
              " '16887895',\n",
              " '11346963',\n",
              " '11117263',\n",
              " '11070970',\n",
              " '8019781']"
            ]
          },
          "metadata": {},
          "execution_count": 20
        }
      ],
      "source": [
        "idlist = data['esearchresult']['idlist']\n",
        "idlist"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# # Define the URL for the NCBI Entrez EFetch API request\n",
        "# fetch_url = \"https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi\"\n",
        "\n",
        "# # Define the parameters for the API request\n",
        "# fetch_params = {\n",
        "#     'db': 'pubmed',\n",
        "#     'retmode': 'xml',\n",
        "#     'id': ','.join(data['esearchresult']['idlist'])  # Join the IDs into a comma-separated string\n",
        "# }\n",
        "\n",
        "# # Send the GET request to the NCBI Entrez API\n",
        "# fetch_response = requests.get(fetch_url, params=fetch_params)\n",
        "\n",
        "# # Check if the request was successful\n",
        "# if fetch_response.status_code == 200:\n",
        "#     # Print the content of the papers\n",
        "#     print(fetch_response.text)\n",
        "# else:\n",
        "#     print(f\"Error: {fetch_response.status_code}\")\n"
      ],
      "metadata": {
        "id": "zQCfYZ_ItEyA"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "import requests"
      ],
      "metadata": {
        "id": "tfBHQQqDPHEm"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# Define the URL for the NCBI Entrez EFetch API request\n",
        "fetch_url = \"https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi\"\n",
        "qa_ans = ''\n",
        "# Loop through all the IDs and fetch their details\n",
        "for paper_id in data['esearchresult']['idlist']:\n",
        "    # Define the parameters for the API request\n",
        "    fetch_params = {\n",
        "        'db': 'gene',\n",
        "        'retmode': 'abstract',\n",
        "        'id': paper_id  # Fetching details for the current ID\n",
        "    }\n",
        "\n",
        "    # Send the GET request to the NCBI Entrez API\n",
        "    fetch_response = requests.get(fetch_url, params=fetch_params)\n",
        "\n",
        "    # Check if the request was successful\n",
        "    if fetch_response.status_code == 200:\n",
        "        # Print the content of the paper\n",
        "        print(f\"Details for ID {paper_id}:\")\n",
        "        print(fetch_response.text)\n",
        "        qa_ans += fetch_response.text\n",
        "    else:\n",
        "        print(f\"Error fetching ID {paper_id}: {fetch_response.status_code}\")\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "nPqM45RMtqCu",
        "outputId": "48c707d7-55b8-4301-c4b9-7891c69ad9a2"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Details for ID 39448407:\n",
            "\n",
            "1.  Error occurred: cannot get document summary\n",
            "ID: 39448407\n",
            "\n",
            "\n",
            "Details for ID 33771198:\n",
            "\n",
            "1.  Error occurred: cannot get document summary\n",
            "ID: 33771198\n",
            "\n",
            "\n",
            "Details for ID 31681346:\n",
            "\n",
            "1. rpmD\n",
            "50S ribosomal protein L30 [Mycobacteroides chelonae CCUG 47445]\n",
            "Other Aliases: BB28_RS19045, BB28_19440\n",
            "Other Designations: 50S ribosomal protein L30\n",
            "ID: 31681346\n",
            "\n",
            "\n",
            "Details for ID 29882727:\n",
            "\n",
            "1. PSLF89_RS23595\n",
            " [Piscirickettsia salmonis LF-89 = ATCC VR-1361]\n",
            "Other Aliases: PSLF89_RS23595, PSLF89_05160\n",
            "This record was discontinued.\n",
            "ID: 29882727\n",
            "\n",
            "\n",
            "Error fetching ID 29666780: 429\n",
            "Error fetching ID 29664328: 429\n",
            "Error fetching ID 28359213: 429\n",
            "Error fetching ID 28298243: 429\n",
            "Error fetching ID 24351075: 429\n",
            "Details for ID 23675379:\n",
            "\n",
            "1. CPC231_RS19750\n",
            "hypothetical protein [Corynebacterium pseudotuberculosis C231]\n",
            "Other Aliases: CPC231_RS19750\n",
            "Other Designations: hypothetical protein\n",
            "Annotation:  NC_017301.1 (2056010..2056948)\n",
            "This record was discontinued.\n",
            "ID: 23675379\n",
            "\n",
            "\n",
            "Details for ID 17478636:\n",
            "\n",
            "1.  Error occurred: cannot get document summary\n",
            "ID: 17478636\n",
            "\n",
            "\n",
            "Details for ID 17467282:\n",
            "\n",
            "1. STP1_0203\n",
            "D-tyrosyl-tRNA(Tyr) deacylase [Staphylococcus pasteuri SP1]\n",
            "Other Aliases: STP1_0203\n",
            "Annotation:  NC_022737.1 (230592..231044, complement)\n",
            "This record was discontinued.\n",
            "ID: 17467282\n",
            "\n",
            "\n",
            "Error fetching ID 16887895: 429\n",
            "Error fetching ID 11346963: 429\n",
            "Error fetching ID 11117263: 429\n",
            "Error fetching ID 11070970: 429\n",
            "Error fetching ID 8019781: 429\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "response = model.generate_content(f\"User Question: {question}\\nContext\\n{qa_ans}\\nAnswer (answer confidently, if you cant find the answer from the provided context, you can generate answer using your existing knowledge base.):\",\n",
        "                                      safety_settings = {\n",
        "                                          HarmCategory.HARM_CATEGORY_HATE_SPEECH: HarmBlockThreshold.BLOCK_NONE,\n",
        "                                          HarmCategory.HARM_CATEGORY_HARASSMENT: HarmBlockThreshold.BLOCK_NONE,\n",
        "                                          HarmCategory.HARM_CATEGORY_SEXUALLY_EXPLICIT: HarmBlockThreshold.BLOCK_NONE,\n",
        "                                          HarmCategory.HARM_CATEGORY_DANGEROUS_CONTENT: HarmBlockThreshold.BLOCK_NONE,\n",
        "                                      })"
      ],
      "metadata": {
        "id": "RMpBgspoz1Us"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "from IPython.display import display, Markdown, Latex\n",
        "display(Markdown(response.text))"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 81
        },
        "id": "hB_NsGLT0Cph",
        "outputId": "21102236-f135-4acc-e6e7-d41b56d00d1f"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.Markdown object>"
            ],
            "text/markdown": "The provided context refers to various gene IDs and protein information, but none of it mentions an \"ids3 chromosome.\"  There's no widely recognized chromosome named \"ids3\" in any common organism.  It's possible this is a typo, a very specific gene name in a niche organism, or related to a specialized database not mentioned here.  Without more information, it's impossible to determine where an \"ids3 chromosome\" would occur.\n"
          },
          "metadata": {}
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "4ZJdQxXYz1SJ"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "WOwc9vLQz1Px"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "Q86rYkbbz1Mj"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "pLWDkwq7z1KM"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "sa4wv0HPz1HZ"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "2HfpkUSTz1Eg"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "D8PcG6rfz1Bi"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 352
        },
        "id": "UVCc7IaN-gzA",
        "collapsed": true,
        "outputId": "dd50a47c-210a-4fd6-824e-c9c771dbf0c7"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi?db=gene&retmax=5&retmode=json&sort=relevance&term=LMP10\n",
            "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=gene&retmax=5&retmode=json&id=19171,5699,8138\n",
            "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esummary.fcgi?db=snp&retmax=10&retmode=json&id=1217074595\n",
            "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi?db=Pubmed&retmax=20&retmode=json&sort=relevance&term=Meesmann+corneal+dystrophy\n",
            "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esummary.fcgi?db=omim&retmax=20&retmode=json&id=618767,601687,300778,148043,122100\n",
            "https://blast.ncbi.nlm.nih.gov/blast/Blast.cgi?CMD=Put&PROGRAM=blastn&MEGABLAST=on&DATABASE=nt&FORMAT_TYPE=XML&QUERY=ATTCTGCCTTTAGTAATTTGATGACAGAGACTTCTTGGGAACCACAGCCAGGGAGCCACCCTTTACTCCACCAACAGGTGGCTTATATCCAATCTGAGAAAGAAAGAAAAAAAAAAAAGTATTTCTCT&HITLIST_SIZE=5\n",
            "https://blast.ncbi.nlm.nih.gov/blast/Blast.cgi?CMD=Get&FORMAT_TYPE=Text&RID=NABS02WF013\n"
          ]
        },
        {
          "output_type": "error",
          "ename": "FileNotFoundError",
          "evalue": "[Errno 2] No such file or directory: 'data/geneturing.json'",
          "traceback": [
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
            "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
            "\u001b[0;32m<ipython-input-25-9b7cc2535d64>\u001b[0m in \u001b[0;36m<cell line: 21>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     19\u001b[0m \u001b[0;31m# initialize\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     20\u001b[0m \u001b[0mprev_call\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtime\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtime\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 21\u001b[0;31m \u001b[0mqas\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mjson\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mload\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mopen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'data/geneturing.json'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     22\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     23\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mtask\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0minfo\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mqas\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mitems\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: 'data/geneturing.json'"
          ]
        }
      ],
      "source": [
        "counter = 1\n",
        "\n",
        "# rough number of chars for truncating\n",
        "# codex accepts 8k tokens ~ 18k chars\n",
        "cut_length = 18000\n",
        "\n",
        "# str_mask is a string of six 0/1 marking whether a in-context learning component is used\n",
        "# six digits correspond to Dc. 1-2, Dm. 1-4\n",
        "\n",
        "# str_mask = sys.argv[1]\n",
        "str_mask = \"111111\"\n",
        "mask = [bool(int(x)) for x in str_mask]\n",
        "prompt = get_prompt_header(mask)\n",
        "\n",
        "# results are saved in the dir of six digits\n",
        "if not os.path.isdir(str_mask):\n",
        "  os.mkdir(str_mask)\n",
        "\n",
        "# initialize\n",
        "prev_call = time.time()\n",
        "qas = json.load(open('data/geneturing.json'))\n",
        "\n",
        "for task, info in qas.items():\n",
        "  if counter <=5:\n",
        "    if os.path.exists(os.path.join(str_mask, f'{task}.json')):\n",
        "      # continue if task is done\n",
        "      preds = json.load(open(os.path.join(str_mask, f'{task}.json')))\n",
        "      if len(preds) == 50: continue\n",
        "\n",
        "    output = []\n",
        "    print(f'Doing task {task}')\n",
        "    for question, answer in info.items():\n",
        "      print('---New Instance---')\n",
        "      question = \"what is ids3?\"\n",
        "      print(question)\n",
        "      q_prompt = prompt + f'Question: {question}\\n'\n",
        "\n",
        "      # save the prompting logs\n",
        "      prompts = []\n",
        "\n",
        "      # record API call times\n",
        "      num_calls = 0\n",
        "\n",
        "      while True:\n",
        "        if len(q_prompt) > cut_length:\n",
        "          # truncate from the start\n",
        "          q_prompt = q_prompt[len(q_prompt) - cut_length:]\n",
        "\n",
        "        # body = {\n",
        "        #   \"model\": \"code-davinci-002\",\n",
        "        #   \"prompt\": q_prompt,\n",
        "        #   \"max_tokens\": 512,\n",
        "        #   \"temperature\": 0,\n",
        "        #   \"stop\": ['->', '\\n\\nQuestion'],\n",
        "        #   \"n\": 1\n",
        "        # }\n",
        "\n",
        "        delta = time.time() - prev_call\n",
        "\n",
        "        # codex has a rate limite of 20 requests / min\n",
        "        # it's a workaround\n",
        "        if delta < 3.1:\n",
        "          time.sleep(3.1 - delta)\n",
        "\n",
        "        try:\n",
        "          prev_call = time.time()\n",
        "          # response = openai.Completion.create(**body)\n",
        "          print(f'User Prompt: {q_prompt}')\n",
        "          response = model.generate_content(q_prompt,\n",
        "                                            safety_settings = {\n",
        "                                                HarmCategory.HARM_CATEGORY_HATE_SPEECH: HarmBlockThreshold.BLOCK_NONE,\n",
        "                                                HarmCategory.HARM_CATEGORY_HARASSMENT: HarmBlockThreshold.BLOCK_NONE,\n",
        "                                                HarmCategory.HARM_CATEGORY_SEXUALLY_EXPLICIT: HarmBlockThreshold.BLOCK_NONE,\n",
        "                                                HarmCategory.HARM_CATEGORY_DANGEROUS_CONTENT: HarmBlockThreshold.BLOCK_NONE,\n",
        "                                            })\n",
        "        # except openai.error.InvalidRequestError:\n",
        "        except:\n",
        "          output.append([question, answer, 'lengthError', prompts])\n",
        "          break\n",
        "\n",
        "        # text = response['choices'][0]['text']\n",
        "        text = response.text\n",
        "        print(text)\n",
        "        num_calls += 1\n",
        "\n",
        "        prompts.append([q_prompt, text])\n",
        "\n",
        "        url_regex = r'\\[(https?://[^\\[\\]]+)\\]'\n",
        "        matches = re.findall(url_regex, text)\n",
        "        if matches:\n",
        "          url = matches[0]\n",
        "\n",
        "          # wait till the BLAST is done on NCBI server\n",
        "          if 'blast' in url and 'Get' in url: time.sleep(30)\n",
        "\n",
        "          call = call_api(url)\n",
        "\n",
        "          if 'blast' in url and 'Put' in url:\n",
        "            rid = re.search('RID = (.*)\\n', call.decode('utf-8')).group(1)\n",
        "            call = rid\n",
        "\n",
        "          if len(call) > 10000:\n",
        "            call = call[:10000]\n",
        "\n",
        "          q_prompt = f'{q_prompt}{text}->[{call}]\\n'\n",
        "\n",
        "        else:\n",
        "          output.append([question, answer, text, prompts])\n",
        "          break\n",
        "\n",
        "        # prevent dead loops\n",
        "        if num_calls >= 10:\n",
        "          output.append([question, answer, 'numError', prompts])\n",
        "          break\n",
        "\n",
        "    counter += 1\n",
        "\n",
        "  with open(os.path.join(str_mask, f'{task}.json'), 'w') as f:\n",
        "    json.dump(output, f, indent=4)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "t3QmVbS8-goR"
      },
      "outputs": [],
      "source": []
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 352
        },
        "id": "HuWKjgUPld2v",
        "outputId": "90b781c6-42df-47c5-a645-cfa09e292813"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi?db=gene&retmax=5&retmode=json&sort=relevance&term=LMP10\n",
            "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=gene&retmax=5&retmode=json&id=19171,5699,8138\n",
            "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esummary.fcgi?db=snp&retmax=10&retmode=json&id=1217074595\n",
            "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi?db=Pubmed&retmax=20&retmode=json&sort=relevance&term=Meesmann+corneal+dystrophy\n",
            "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esummary.fcgi?db=omim&retmax=20&retmode=json&id=618767,601687,300778,148043,122100\n",
            "https://blast.ncbi.nlm.nih.gov/blast/Blast.cgi?CMD=Put&PROGRAM=blastn&MEGABLAST=on&DATABASE=nt&FORMAT_TYPE=XML&QUERY=ATTCTGCCTTTAGTAATTTGATGACAGAGACTTCTTGGGAACCACAGCCAGGGAGCCACCCTTTACTCCACCAACAGGTGGCTTATATCCAATCTGAGAAAGAAAGAAAAAAAAAAAAGTATTTCTCT&HITLIST_SIZE=5\n",
            "https://blast.ncbi.nlm.nih.gov/blast/Blast.cgi?CMD=Get&FORMAT_TYPE=Text&RID=NABU1HCY013\n"
          ]
        },
        {
          "output_type": "error",
          "ename": "FileNotFoundError",
          "evalue": "[Errno 2] No such file or directory: 'data/geneturing.json'",
          "traceback": [
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
            "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
            "\u001b[0;32m<ipython-input-26-41a13d8c2342>\u001b[0m in \u001b[0;36m<cell line: 19>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     17\u001b[0m \u001b[0;31m# initialize\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     18\u001b[0m \u001b[0mprev_call\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtime\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtime\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 19\u001b[0;31m \u001b[0mqas\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mjson\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mload\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mopen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'data/geneturing.json'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     20\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     21\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mtask\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0minfo\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mqas\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mitems\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: 'data/geneturing.json'"
          ]
        }
      ],
      "source": [
        "# rough number of chars for truncating\n",
        "# codex accepts 8k tokens ~ 18k chars\n",
        "cut_length = 18000\n",
        "\n",
        "# str_mask is a string of six 0/1 marking whether a in-context learning component is used\n",
        "# six digits correspond to Dc. 1-2, Dm. 1-4\n",
        "\n",
        "# str_mask = sys.argv[1]\n",
        "str_mask = \"111111\"\n",
        "mask = [bool(int(x)) for x in str_mask]\n",
        "prompt = get_prompt_header(mask)\n",
        "\n",
        "# results are saved in the dir of six digits\n",
        "if not os.path.isdir(str_mask):\n",
        "  os.mkdir(str_mask)\n",
        "\n",
        "# initialize\n",
        "prev_call = time.time()\n",
        "qas = json.load(open('data/geneturing.json'))\n",
        "\n",
        "for task, info in qas.items():\n",
        "  if os.path.exists(os.path.join(str_mask, f'{task}.json')):\n",
        "    # continue if task is done\n",
        "    preds = json.load(open(os.path.join(str_mask, f'{task}.json')))\n",
        "    if len(preds) == 50: continue\n",
        "\n",
        "  output = []\n",
        "  print(f'Doing task {task}')\n",
        "  for question, answer in info.items():\n",
        "    print('---New Instance---')\n",
        "    question = \"what is ids3?\"\n",
        "    print(question)\n",
        "    q_prompt = prompt + f'Question: {question}\\n'\n",
        "\n",
        "    # save the prompting logs\n",
        "    prompts = []\n",
        "\n",
        "    # record API call times\n",
        "    num_calls = 0\n",
        "\n",
        "    while True:\n",
        "      if len(q_prompt) > cut_length:\n",
        "        # truncate from the start\n",
        "        q_prompt = q_prompt[len(q_prompt) - cut_length:]\n",
        "\n",
        "      # body = {\n",
        "      #   \"model\": \"code-davinci-002\",\n",
        "      #   \"prompt\": q_prompt,\n",
        "      #   \"max_tokens\": 512,\n",
        "      #   \"temperature\": 0,\n",
        "      #   \"stop\": ['->', '\\n\\nQuestion'],\n",
        "      #   \"n\": 1\n",
        "      # }\n",
        "\n",
        "      delta = time.time() - prev_call\n",
        "\n",
        "      # codex has a rate limite of 20 requests / min\n",
        "      # it's a workaround\n",
        "      if delta < 3.1:\n",
        "        time.sleep(3.1 - delta)\n",
        "\n",
        "      try:\n",
        "        prev_call = time.time()\n",
        "        # response = openai.Completion.create(**body)\n",
        "        print(f'User Prompt: {q_prompt}')\n",
        "        response = model.generate_content(q_prompt,\n",
        "                                          safety_settings = {\n",
        "                                              HarmCategory.HARM_CATEGORY_HATE_SPEECH: HarmBlockThreshold.BLOCK_NONE,\n",
        "                                              HarmCategory.HARM_CATEGORY_HARASSMENT: HarmBlockThreshold.BLOCK_NONE,\n",
        "                                              HarmCategory.HARM_CATEGORY_SEXUALLY_EXPLICIT: HarmBlockThreshold.BLOCK_NONE,\n",
        "                                              HarmCategory.HARM_CATEGORY_DANGEROUS_CONTENT: HarmBlockThreshold.BLOCK_NONE,\n",
        "                                          })\n",
        "      # except openai.error.InvalidRequestError:\n",
        "      except:\n",
        "        output.append([question, answer, 'lengthError', prompts])\n",
        "        break\n",
        "\n",
        "      # text = response['choices'][0]['text']\n",
        "      text = response.text\n",
        "      print(text)\n",
        "      num_calls += 1\n",
        "\n",
        "      prompts.append([q_prompt, text])\n",
        "\n",
        "      url_regex = r'\\[(https?://[^\\[\\]]+)\\]'\n",
        "      matches = re.findall(url_regex, text)\n",
        "      if matches:\n",
        "        url = matches[0]\n",
        "\n",
        "        # wait till the BLAST is done on NCBI server\n",
        "        if 'blast' in url and 'Get' in url: time.sleep(30)\n",
        "\n",
        "        call = call_api(url)\n",
        "\n",
        "        if 'blast' in url and 'Put' in url:\n",
        "          rid = re.search('RID = (.*)\\n', call.decode('utf-8')).group(1)\n",
        "          call = rid\n",
        "\n",
        "        if len(call) > 10000:\n",
        "          call = call[:10000]\n",
        "\n",
        "        q_prompt = f'{q_prompt}{text}->[{call}]\\n'\n",
        "\n",
        "      else:\n",
        "        output.append([question, answer, text, prompts])\n",
        "        break\n",
        "\n",
        "      # prevent dead loops\n",
        "      if num_calls >= 10:\n",
        "        output.append([question, answer, 'numError', prompts])\n",
        "        break\n",
        "\n",
        "  with open(os.path.join(str_mask, f'{task}.json'), 'w') as f:\n",
        "    json.dump(output, f, indent=4)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "6GUHjkItQyGa"
      },
      "outputs": [],
      "source": []
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "lUaYK3pJQyDZ"
      },
      "outputs": [],
      "source": []
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "-QB9--gkQyBP"
      },
      "outputs": [],
      "source": []
    }
  ],
  "metadata": {
    "colab": {
      "provenance": [],
      "include_colab_link": true
    },
    "kernelspec": {
      "display_name": "Python 3",
      "name": "python3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}